src/firefly/js/drawingLayers/WebGrid.js

```diff
@@ -0,0 +1,882 @@
+         Put the label in the middle of this direction.
+         */
+        var vM=Math.trunc(xLines[nLevel0].length/2);
+        for (let i=0; i< nLevel1; i++){
+            points[i+ nLevel0] = makeImageWorkSpacePt(xLines[nLevel0 +i][vM], yLines[nLevel0 +i][vM]);
+        }
+    }
+    return  points;
+}
+
+
+function drawLabeledPolyLine (drawData, bounds, nLevel0,  label, labelLocations, x, y, count, aitoff,screenWidth){
+
+
+    //add the  draw line data to the drawData
+    var ipt0, ipt1;
+    for (let i=0; i<x.length-1; i+=1) {
+        //check the x[i] and y[i] are inside the image screen
+        if (x[i] > -1000 && x[i+1] > -1000 &&
+            ((x[i] >= bounds.x) &&
+            ((x[i] - bounds.x) < bounds.width) &&
+            (y[i] >= bounds.y) &&
+            ((y[i]-bounds.y) < bounds.height) ||
+                // bounds check on x[i+1], y[i+1]
+            (x[i+1] >= bounds.x) &&
+            ((x[i+1] - bounds.x) < bounds.width) &&
+            (y[i+1] >= bounds.y) &&
+            ((y[i+1]-bounds.y) < bounds.height))) {
+            ipt0= makeImageWorkSpacePt(x[i],y[i]);
+            ipt1= makeImageWorkSpacePt(x[i+1], y[i+1]);
+            if (!aitoff  ||  ((Math.abs(ipt1.x-ipt0.x) <screenWidth /8 ) && (aitoff))) {
+                drawData.push(ShapeDataObj.makeLine(ipt0,ipt1));
+            }
+        } //if
+    } // for
+
+    // draw the label.
+    if (useLabels){
+        if (count< nLevel0) { //vertical line labels
+            drawData.push(ShapeDataObj.makeText(labelLocations[count], label));
+        }
+        else { //Horizontal line labels
+            drawData.push(ShapeDataObj.makeText(labelLocations[count], label));
+        }
+    }
+}
+
+function drawLines(csys,bounds, labels, xLines,yLines, nLevel0, nLevel1, aitoff,screenWidth) {
+    // Draw the lines previously computed.
+    //get the locations where to put the labels
+    var labelLocations =getLabelPoints(bounds, csys, xLines, yLines, nLevel0, nLevel1);
+    var drawData=[];
+    var  lineCount = xLines.length;
+    for (let i=0; i<lineCount; i+=1) {
+        drawLabeledPolyLine(drawData, bounds,nLevel0, labels[i],labelLocations ,
+            xLines[i], yLines[i], i, aitoff,screenWidth);
+    }
+    return drawData;
+
+}
+
+
+function computeLines(cc, csys, range, levels,screenWidth) {
+    /* This is where we do all the work. */
+    /* range and levels have a first dimension indicating x or y
+     * and a second dimension for the different values (2 for range)
+     * and a possibly variable number for levels.
+     */
+
+    var xLines = [];
+    var yLines = [];
+    var offset = 0;
+    var points=[];
+    for (let i=0; i<2; i++) {
+        for (let j=0; j<levels[i].length; j++) {
+            points = findLine(cc, csys,i, levels[i][j], range,screenWidth);
+            xLines[offset] = points[0];
+            yLines[offset] = points[1];
+            offset += 1;
+        }
+    }
+    return {xLines, yLines};
+}
+
+
+
```

Comment:
I think you need to break out the drawing layer construction into another file with the entry point of `getData(plot, cc)`. That way the it is easier for focus only how the drawing layer is handled vs. how the grid is constructed.